### PR TITLE
[csharp] Add scope for oauth2

### PR DESCRIPTION
**Pull request by mjansrud**
[csharp] Add scope for oauth2 based on issue https://github.com/OpenAPITools/openapi-generator/issues/17047. 

When using the OAuth client credentials flow, the Configuration object allows me to specify the OAuthTokenUrl, OAuthClientId, and OAuthClientSecret, but there seems to be no way to specify any scope. Problem I'm facing is that I need to specify a scope to retrieve a token, so currently the request made in the OAuthAuthenticator.cs (GetToken() method) class returns a 400 response.

Microsoft identity platform documentation:
https://learn.microsoft.com/en-us/entra/identity-platform/v2-oauth2-client-creds-grant-flow

image

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh ./bin/configs/*.yaml
  ./bin/utils/export_docs_generators.sh
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming 7.6.0 minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

@iBicha

### CODE REVIEW

**Comment by wing328** [comment]
thanks for the PR. can you please review the test failures when you've time?

**Comment by mjansrud** [comment]
I added LangVersion 8 to enable nullable strings in the `netstandard2.0` sample projects to fix the tests. Im not sure if this is a ok solution, or if we want to solve the issue in another way. What do you think @wing328 ?

**Comment by wing328** [comment]
> I added LangVersion 8 to enable nullable strings

can it be avoided if possible?

**Comment by mjansrud** [comment]
I think i found a decent solution by only setting the string as nullable on supported frameworks with templating rules. It seems like the variable can be set as a regular string in netstandard2.0, but still be defined as null. See the image screenshot attached. Can you run the rest of the tests? :)  

image

**Comment by mjansrud** [comment]
Ive updated the samples @wing328

File: modules/openapi-generator/src/main/resources/csharp/Configuration.mustache [hunk 1]
```diff
@@ -366,6 +366,12 @@ namespace {{packageName}}.Client
         /// <value>The OAuth Client Secret.</value>
         public virtual string OAuthClientSecret { get; set; }
 
+        /// <summary>
+        /// Gets or sets the client scope for OAuth2 authentication.
+        /// </summary>
+        /// <value>The OAuth Client Scope.</value>
+        public virtual string{{^netStandard}}?{{/netStandard}} OAuthScope { get; set; }
```
Comment (devhl-labs) [inline modules/openapi-generator/src/main/resources/csharp/Configuration.mustache:373]: Please use {{nrt?}}

File: modules/openapi-generator/src/main/resources/csharp/IReadableConfiguration.mustache [hunk 2]
```diff
@@ -43,6 +43,12 @@ namespace {{packageName}}.Client
         /// <value>OAuth Client Secret.</value>
         string OAuthClientSecret { get; }
 
+        /// <summary>
+        /// Gets the OAuth token scope.
+        /// </summary>
+        /// <value>OAuth Token scope.</value>
+        string{{^netStandard}}?{{/netStandard}} OAuthScope { get; }
```
Comment (devhl-labs) [inline modules/openapi-generator/src/main/resources/csharp/IReadableConfiguration.mustache:50]: Please use {{nrt?}}

**Comment by mjansrud** [comment]
Ive changed to {{nrt?}} as requested @devhl-labs. Does it look ok? And can one of you please generate the tests for me? That would be much appreciated @wing328 @devhl-labs. Upon test generation i get a deviation from main, and have to change the code manually, which is a real pain.

image

image

**Review by devhl-labs — APPROVED** [review]

**Comment by mjansrud** [comment]
@wing328 can you run the last tests and take a final look?

**Comment by wing328** [comment]
https://github.com/OpenAPITools/openapi-generator/actions/runs/10005218697/job/27738318490?pr=19175

can you please update the samples?

**Comment by mjansrud** [comment]
Hi Wing. Did you see comment https://github.com/OpenAPITools/openapi-generator/pull/19175#issuecomment-2238691237?

**Comment by wing328** [comment]
filed https://github.com/OpenAPITools/openapi-generator/pull/19234 with updated samples. let's see how that goes

**Comment by wing328** [comment]
@mjansrud thanks for the PR, which has been merged via #19234 with updated samples (all tests passed)

**Pull request closed without merging**